### ISSUE TITLE
A potential bug for multi-GPU training

### ISSUE THREAD

**Issue by zyushun** [Lightning-AI/litgpt#1368]
Hi,

I found the following strange phenomena when running your code for tinyllama pretraining. 

1. When using multiple GPUs, I got **completely different results** when **running the same code twice**.  Further, many loss spike occurs. See the example for 2-card training. I use all the default settings except that I shrink the learning rate from 4e-4 to 2e-4 and batchsize from 1024 to 512. 

AdamW 2-card: run1

wandb: 🚀 View run at https://wandb.ai/yushunzhang0410/pretrain-tiny-llama-1.1b/runs/83b8yfjz

AdamW 2-card: run2

wandb: 🚀 View run at https://wandb.ai/yushunzhang0410/pretrain-tiny-llama-1.1b/runs/8p6axrgw

Two runs are totally different and the training fails.

2. When simply changing the above settings to single GPU, these issues do not occur. Two runs are mostly the same (with slight difference though) and the loss decreases stably without any spikes.

AdamW 1-card: run 1

wandb: 🚀 View run at https://wandb.ai/yushunzhang0410/pretrain-tiny-llama-1.1b/runs/kdg2qmj8

AdamW 1-card: run 2 

wandb: 🚀 View run at https://wandb.ai/yushunzhang0410/pretrain-tiny-llama-1.1b/runs/vh23qd0u

Two runs are mostly the same and the loss decreases stably.

Do you encounter a similar issue? Any idea why?

**Comment by lantiga** [comment]
Thanks for the report. Can you try:

- running without torch.compile (comment this line out https://github.com/Lightning-AI/litgpt/blob/main/litgpt/pretrain.py#L174)
- running with torch.compile but on PyTorch 2.3

Thanks a lot for investigating this.

cc @awaelchli for visibility

**Comment by zyushun** [comment]
Hi,

I still encounter this issue when using **your latest code on github**. 

four A800-80GB GPU, AdamW, Tinyllama, all default settings. I did not change anything except the data path. I still encounter loss spike which does not exists in single-GPU training.  

wandb: 🚀 View run at https://wandb.ai/yushunzhang0410/pretrain-tiny-llama-1.1b-litgpt-version/runs/bhiopo5z

<img width="548" alt="image" src="https://github.com/Lightning-AI/litgpt/assets/85537560/138a5ec4-c778-40fb-95b1-222f6f0ea0b9">


 I simply use pip install 'litgpt[all]' to get all the dependencies, as you suggested in the github. I checked your default pretrain.py and find I am using model.compile, with Pytorch 2.3.0. This meets your suggestion "running with torch.compile but on PyTorch 2.3"
 
 What should I do now? Am I the only one encountering this issue?  Do you have this issue on your side? I think you can easily reproduce this issue if you git clone +  pip install 'litgpt[all]'  +  run the code (just as I did).

**Comment by awaelchli** [comment]
From your WandB logs I can tell that you are using lightning 2.2.0dev. Please upgrade this to the latest version (which is 2.2.3) and double check that your code runs with this version (because it contains [this FSDP fix](https://github.com/Lightning-AI/pytorch-lightning/pull/19755)). 

I can also tell that you've made modifications to the code. This of course makes it impossible for us to know what is going on. Before you make modifications, you should run with our version. Please commit your current state to a branch, then 

```
git checkout main
git pull
pip install -e ".[all]"
pip install -U lightning torch
```
in a fresh environment. Then continue with the instructions for preparing the data [here](https://github.com/Lightning-AI/litgpt/blob/main/tutorials/pretrain_tinyllama.md).

And then you can run `litgpt pretrain --config config_hub/pretrain/tinyllama.yaml`. 

The settings in the configs are only valid for the code that we publish. If modifications are made otherwise, there are no guarantees.

**Comment by awaelchli** [comment]
Your wandb log metadata suggests you are using lightning 2.2dev, which probably came with an older version of litgpt that you had. You might need [this fix](https://github.com/Lightning-AI/pytorch-lightning/pull/19755) for pretraining, so I suggest updating lightning to the latest version as the immedate action here.

**Comment by zyushun** [comment]
Hi,

Thanks for your prompt reply.

I think my lightning version is correct. My code is based on a fresh environment created yesterday, where I simply run "  pip install 'litgpt[all]'  "as you suggested in the github.  As confirmed in the "conda list" screenshot below,  I am using lightning 2.3.0.dev20240328 . 

<img width="888" alt="image" src="https://github.com/Lightning-AI/litgpt/assets/85537560/924dd510-47ea-4c7b-b825-b0911f3e91de">

Any other possible issue?

**Comment by awaelchli** [comment]
The initialization fix I made was on April 11, so the package you have is still too old. The fix was then cherry-picked into lightning 2.2.2. So I would still update the package.
